core/io/set_encoding_by_bom_spec.rb

```diff
@@ -0,0 +1,42 @@
+require_relative '../../spec_helper'
+
+describe "IO#set_encoding_by_bom" do
+  before :each do
+    @name = tmp('io_set_encoding_by_bom.txt')
+    touch(@name)
+    @io = new_io(@name, 'rb')
+  end
+
+  after :each do
+    @io.close unless @io.closed?
+    rm_r @name
+  end
+
+  ruby_version_is "2.7" do
+    it "returns the result encoding if found BOM UTF-8 sequence" do
+      File.write(@name, "\u{FEFF}abc")
+
+      @io.set_encoding_by_bom.should == Encoding::UTF_8
+      @io.external_encoding.should == Encoding::UTF_8
+    end
+
+    it "returns nil if found BOM sequence not provided" do
+      File.write(@name, "abc")
+
+      @io.set_encoding_by_bom.should == nil
+    end
+
+    it 'returns exception if io not in binary mode' do
+      not_binary_io = new_io(@name, 'r')
+
+      -> { not_binary_io.set_encoding_by_bom }.should raise_error(ArgumentError, 'ASCII incompatible encoding needs binmode')
+      not_binary_io.close
```

Comment:
Could you move this in a begin/ensure/end? So even if the assertion fails, the IO is still closed.